add auth header (#127)
dgraph-io/dgraph-docs: content/graphql/lambda/overview.md

    name = "Overview"
+++

Lambda provides a way to write your custom logic in JavaScript, integrate it
with your GraphQL schema, and execute it using the GraphQL API in a few easy
steps:

- Setup a Dgraph cluster with a working lambda server (not required for
  [Slash GraphQL](https://dgraph.io/slash-graphql) users)
- Declare lambda queries, mutations, and fields in your GraphQL schema as needed
- Define lambda resolvers for them in a JavaScript file

This also simplifies the job of developers, as they can build a complex backend
that is rich with business logic, without setting up multiple different
services. Also, you can build your backend in JavaScript, which means you can
build both your frontend and backend using the same language.

Dgraph doesn't execute your custom logic itself. It makes external HTTP requests
to a user-defined lambda server.

> [Slash GraphQL](https://dgraph.io/slash-graphql) will do all of this for you.

Reviewer: ```suggestion
[Dgraph Cloud](https://dgraph.io/cloud) will do all of this for you.
```

{{% notice "tip" %}} If you want to deploy your own lambda server, you can find
the implementation of Dgraph Lambda in our
[open-source repository](https://github.com/dgraph-io/dgraph-lambda). Please
refer to the documentation on
[setting up a lambda server](/graphql/lambda/server) for more details.
{{% /notice %}}

{{% notice "note" %}} If you're using

> [Slash GraphQL](https://dgraph.io/slash-graphql), the final compiled script file

Reviewer: ```suggestion
[Dgraph Cloud](https://dgraph.io/cloud), the final compiled script file
```

must be under 500Kb {{% /notice %}}

## Declaring lambda in a GraphQL schema

> There are three places where you can use the `@lambda` directive and thus tell

Reviewer: and thus tell -> to tell

## Defining lambda resolvers in JavaScript

A lambda resolver is a user-defined JavaScript function that performs custom
actions over the GraphQL types, interfaces, queries, and mutations. There are
two methods to register JavaScript resolvers:

- `self.addGraphQLResolvers`
- `self.addMultiParentGraphQLResolvers`

> {{% notice "tip" %}} Functions `self.addGraphQLResolvers` and

Reviewer: ```suggestion
{{% notice "tip" %}} The functions `self.addGraphQLResolvers` and
```

  function, so this can be used securely
- `dql`, provides an API to execute DQL from the lambda server
- `authHeader`, provides the JWT key and value of the auth header passed from
  the client

The `addGraphQLResolvers` can be represented with the following TypeScript
types:

```TypeScript
type GraphQLResponse {
  data?: Record<string, any>
  errors?: { message: string }[]
}

type AuthHeader {
  key: string
  value: string
}

type GraphQLEventWithParent = {
  parent: Record<string, any> | null
  args: Record<string, any>
  graphql: (query: string, vars?: Record<string, any>, authHeader?: AuthHeader) => Promise<GraphQLResponse>
  dql: {
    query: (dql: string, vars?: Record<string, any>) => Promise<GraphQLResponse>
    mutate: (dql: string) => Promise<GraphQLResponse>
  }
  authHeader: AuthHeader
}

function addGraphQLResolvers(resolvers: {
  [key: string]: (e: GraphQLEventWithParent) => any;
}): void
```

{{% notice "tip" %}} `self.addGraphQLResolvers` is the default choice for
registering resolvers when the result of the lambda for each parent is
independent of other parents. {{% /notice %}}

Each resolver function should return data in the exact format as the return type

> of GraphQL field, query, or mutation for which it is being registered.

Reviewer: ```suggestion
of the GraphQL field, query, or mutation for which it is being registered.
```

In the following example, the resolver function `myTypeResolver` registered for
the `customField` field in `MyType` returns a string because the return type of
that field in the GraphQL schema is `String`:

```javascript
const myTypeResolver = ({ parent: { customField } }) =>
  `My value is ${customField}.`;

self.addGraphQLResolvers({
  "MyType.customField": myTypeResolver,
});
```

Another resolver example using a `graphql` call:

```javascript
async function todoTitles({ graphql }) {
  const results = await graphql("{ queryTodo { title } }");
  return results.data.queryTodo.map((t) => t.title);
}

self.addGraphQLResolvers({
  "Query.todoTitles": todoTitles,
});
```

### addMultiParentGraphQLResolvers

The `self.addMultiParentGraphQLResolvers` is useful in scenarios where you want

> to perform computations involving all the parents returned from Dgraph for a

Reviewer: ```suggestion
to perform computations involving all of the parents returned from Dgraph for a
```
